File: benchmarks/pandas/bench_pct_change_fn.py
"""
Benchmark: pandas pct_change on Series and DataFrame.
Outputs JSON: {"function": "pct_change_fn", "mean_ms": ..., "iterations": ..., "total_ms": ...}
"""
import json
import time
import pandas as pd

ROWS = 100_000
WARMUP = 3
ITERATIONS = 10

data = [i * 1.1 + 1.0 for i in range(ROWS)]
s = pd.Series(data)
df = pd.DataFrame({"a": data, "b": [x * 2 for x in data]})

for _ in range(WARMUP):
    s.pct_change()
    s.pct_change(periods=2)
    df.pct_change()

times = []
for _ in range(ITERATIONS):
    t0 = time.perf_counter()
    s.pct_change()
    s.pct_change(periods=2)
    df.pct_change()
    times.append((time.perf_counter() - t0) * 1000)

total_ms = sum(times)
mean_ms = total_ms / ITERATIONS
print(json.dumps({"function": "pct_change_fn", "mean_ms": round(mean_ms, 3), "iterations": ITERATIONS, "total_ms": round(total_ms, 3)}))

File: benchmarks/tsb/bench_pct_change_fn.ts
/**
 * Benchmark: pctChangeSeries / pctChangeDataFrame — standalone functional pct_change.
 * Outputs JSON: {"function": "pct_change_fn", "mean_ms": ..., "iterations": ..., "total_ms": ...}
 */
import { Series, DataFrame, pctChangeSeries, pctChangeDataFrame } from "../../src/index.ts";

const ROWS = 100_000;
const WARMUP = 3;
const ITERATIONS = 10;

const data = Array.from({ length: ROWS }, (_, i) => i * 1.1 + 1.0);
const s = new Series({ data });
const df = new DataFrame(
  new Map([
    ["a", new Series({ data })],
    ["b", new Series({ data: data.map((x) => x * 2) })],
  ]),
);

for (let i = 0; i < WARMUP; i++) {
  pctChangeSeries(s);
  pctChangeSeries(s, { periods: 2 });
  pctChangeDataFrame(df);
}

const times: number[] = [];
for (let i = 0; i < ITERATIONS; i++) {
  const t0 = performance.now();
  pctChangeSeries(s);
  pctChangeSeries(s, { periods: 2 });
  pctChangeDataFrame(df);
  times.push(performance.now() - t0);
}

const totalMs = times.reduce((a, b) => a + b, 0);
const meanMs = totalMs / ITERATIONS;
console.log(
  JSON.stringify({
    function: "pct_change_fn",
    mean_ms: Math.round(meanMs * 1000) / 1000,
    iterations: ITERATIONS,
    total_ms: Math.round(totalMs * 1000) / 1000,
  }),
);
